README.rst
============
Jock MKT SDK
============

.. image:: https://img.shields.io/pypi/v/jockmkt-sdk?style=for-the-badge
  :target: https://pypi.python.org/pypi/jockmkt-sdk

.. image:: https://img.shields.io/github/license/nysugfx/jockmkt-sdk?style=for-the-badge
  :target: https://github.com/nysugfx/jockmkt-sdk/LICENSE.txt

.. image:: https://img.shields.io/pypi/pyversions/jockmkt-sdk?style=for-the-badge
  :target: https://pypi.python.org/pypi/jockmkt-sdk



This is an unofficial wrapper built to interface with the Jock MKT API.

*The developer of this sdk is in no way affiliated with Jock MKT -- please use at your own risk.*

The aim of this SDK is to build a host of easy-to-use tools for users to interact with the Jock MKT API. It includes every possible api call, every optional argument, and objects that organize all of their responses. In the future, additions will be made to further simplify these api calls with the addition of useful tools.

`JockMKT Official API Docs <https://docs.jockmkt.com/>`_

**Code snippet**

.. code-block:: python

    # the following code will buy $50 worth of shares for every player in an event
    # whose last traded price is less than the estimated price
    client = Client()
    client.get_auth_token(secret_key, api_key)
    event = client.get_event(event_id, include_tradeables=True)
    for tradeable in event.tradeables:
         if tradeable.last < tradeable.estimated:
             auth.place_order(tradeable.id, price=tradeable.estimated, size=50)


Installation (via PyPi)
-----------------------
.. code-block:: bash

    pip install jockmkt-sdk

`PyPi Link <https://pypi.org/project/jockmkt-sdk/0.1.2/>`_

Note: If you have problems installing, please ensure that you have updated pip and setuptools. If you still have trouble, please install requirements manually.

Contribute
-----------
  - Source: https://github.com/nysugfx/jockmkt-sdk
  - Issue Tracker: https://github.com/nysugfx/jockmkt-sdk/issues

Docs
----
  - `ReadTheDocs <https://jockmkt-sdk.readthedocs.io/en/latest/index.html>`_

Features
--------
- Calls for every API endpoint
- Exception handling
- Simple auth token generation & use
- Simple order placement


Getting Started
===============

Obviously, create an account at `jockmkt.com <jockmkt.com>`_

Contact developers@jockmkt.com for your auth keys

Authorization & basic calls:
  .. code:: python

    from jockmkt-sdk.client import Client
    api_key = '<jm_key_xxx>'
    secret_key = '<xxx>'

    client = Client(secret_key, api_key)

    #display the first 100 nba entities
    players = client.get_entities(start=0, league='nba')

    #get 25 recent and upcoming events:
    events = client.get_events()

    #get the last 500 events:
    fetchall_events = []
    for i in range(5):
         fetchall_events.append(client.get_events(start=i, limit=100))

    #get a single event:
    event_id = events[0].event_id
    event = client.get_event(event_id, include_tradeables=True)

    #join an event:
    entry = client.create_entry(event_id)

    #view event profit:
    print(entry.profit)

    #view event tradeables:
    tradeables = event.tradeables

    #pick a specific player and place an order during IPO phase at the Jock MKT estimated price:
    for player in tradeables:
         if player.name == 'Paul George':
              client.place_order(player.tradeable_id, player.estimated)

    #buy a specified $ amount worth of a player at the market price, during live trading:
    order = client.create_order(id=player.tradeable_id, price=player.ask,
         phase='live', order_size=100)

    #cancel order
    client.cancel_order(order['order']['id'])

Contact
-------
nysu.gfx@gmail.com


changelog.rst
=======================
Changelog - jockmkt-sdk
=======================

This changelog will document all changes after and including release **0.2.0** of the Jockmkt-sdk API Wrapper.

Releases follow `semantic versioning <https://semver.org/spec/v2.0.0.html>`_.
The format is based on `Keep a Changelog <https://keepachangelog.com/en/1.0.0/>`_.

Docs available at: `Read The Docs <https://jockmkt-sdk.readthedocs.io/en/latest/>`_


Release 0.2.15
##############

``ADDED:``

- Option to include object count in your request.
    - If you are trying to get all of your orders, for example, you can set ``include_count=True`` in your request to return the total number of orders. Assists with pagination.
    - e.g.:

.. code-block:: python

    The perfect number of requests to get all orders
    orders, order_count = client.get_orders(include_count=True)
    for i in range(order_count//100 + 1):
        orders.extend(client.get_orders(start=i)

- verbosity feature to get rid of print statements

``FIXED:``

- Typo in Entry docs
- Better self-documentation by renaming some variables within functions. Will not affect return values.
- Fixed key problem in NHL entity

Release 0.2.14
##############

``FIXED:``

- Deleted an annoying print statement.

Release 0.2.13
##############

``FIXED:``

- Typing issue

Release 0.2.12
##############

``FIXED``

- File structure problem

Release 0.2.11
##############

``FIXED``

- No longer raises an exception if the user has already joined an event.
- Now actually returns horse racing entities

``ADDED``

- New websocket connect method that should allow for a more graceful exit.

Release 0.2.10
##############

``FIXED:``

- Notifications from websockets not being pushed to messages
- Websocket messages not being converted to objects correctly

Release 0.2.9
#############

``FIXED:``

- Small incompatibility bug with | in typing

Release 0.2.8
#############

``ADDED:``

- Compatibility with earlier versions of python (>=3.7)

``FIXED:``

- Missing requirements in requirements.txt and pyproject.toml

Release 0.2.7
#############

``ADDED:``

- Added current_shares attribute to Event object
- ssl certification for websocket feed

0.2.6
#####

- added sold count to position object

Release 0.2.3 and 0.2.4
#############

``FIXED:``

- Decimal rounding error for pricing

- Github Actions


Release 0.2.2
#############

``Added:``

- Order response is now an Order object.
    - added direction and time_in_force instance variables

- Balance object

- 'insufficient_funds' and 'mixed_position' message in exception handling

- 'updated_at' attribute for Tradeable objects

``Fixed:``

- Order prices are now formatted using Decimal rather than ``"{0:.2f}".format()`` which was causing the occasional bug.

- ``.place_order()`` input typing was corrected

- Order object is correctly parsed from 'account' websocket endpoint

- kwargs are correctly unpacked in websocket error_handler

``Changed:``

- place_order args adjusted order of positional args:
    - tradeable_id, price, **qty: int = 1**, ...)
    - should not break any code

- Orders can now be placed like so (you do not need qty as a keyword argument):

```
client.place_order('tradeable_id', price, qty)
```


Release 0.2.1
#############

``Fixed:``

- Docs

- Event endpoint displaying only 20 results

Release 0.2.0
#############

``Added:``

- Functionality for websockets

``Fixed:``

- ``get_game_logs`` is now fully functional

- ``get_game_logs`` now includes ``statistics`` and ``projected_statistics`` attributes.

- order rate limit handling is fixed -- no order deletions will count towards the rate limit

Release 0.1.0
#############

``Added:``

- functionality for ALL Jockmkt API endpoints

- Automatically fetches an authorization token

- Objects for every relevant request: (i.e. tradeable, entity, event, order, etc.) that contain attributes for all available information.

- Rate limit handling for orders -- the user can place as many orders as they want and the SDK will automatically cache requests if they hit the limit

- Testing

- Event scoring information

- Full docstrings explaining every api request, return value and instance variable

- Full documentation with examples `here <https://jockmkt-sdk.readthedocs.io/en/latest/>`_








src/basic_api_calls.py
import requests
import json
from exception import JockAPIException
from objects import Team, Game, GameLog, Event, Tradeable, Entry, Order, Position, AccountActivity, Entity, \
    _case_switch_ent


class Client(object):
    """
    Client object automatically obtains an auth token & user balance using the user provided secret & api keys

    :param secret: the user's 32 digit long secret key
    :type secret: str

    :param api_key: the user's api_key, starting with jm_api_
    :type api_key: str
    """
    API_VERSION = 'v1'
    BASE_URL = 'https://api.jockmkt.net'
    _AUTH_TOKEN = {}

    def __init__(self, secret, api_key, request_params=None):
        self.secret = secret
        self.api_key = api_key
        self._request_params = request_params
        _AUTH_TOKEN = self._get_auth_token()
        self.balance = self._get_account_bal()
        self.account = self._get_account()

    def _create_path(self, path, api_version=None):
        """
        generates a path for self._request
        """
        api_version = api_version or self.API_VERSION
        return '/{}/{}'.format(api_version, path)

    def _get_auth_token(self):
        """
        obtains the user's auth token using provided api and secret keys, generating a BearerAuth header
        """
        payload = {
            'grant_type': 'client_credentials',
            'key': str(self.api_key),
            'secret': str(self.secret)
        }
        response = requests.post(f'{self.BASE_URL}/{self.API_VERSION}/oauth/tokens', data=payload).json()
        if response['status'] == 'error':
            raise KeyError("Your authorization keys are not valid!")
        self.token = {'Authorization': 'Bearer ' + response['token']['access_token']}
        return self.token

    def _request(self, method, path, api_version=None, **kwargs) -> json:
        """
        method by which all requests are made
        """
        if self._request_params:
            kwargs.update(self._request_params)
        kwargs['data'] = kwargs.get('data', {})
        kwargs['params'] = kwargs.get('params', {})

        full_path = self._create_path(path, api_version)

        if method == 'get':
            if kwargs['data']:
                kwargs['payload'] = kwargs['data']
            elif kwargs['params']:
                kwargs['payload'] = kwargs['params']
            else:
                kwargs['payload'] = {}
            response = requests.get('{}{}'.format(self.BASE_URL, full_path), params=kwargs['payload'],
                                    headers=self.token)

        if method == 'post':
            if kwargs['data']:
                kwargs['payload'] = kwargs['data']
            elif kwargs['params']:
                kwargs['payload'] = kwargs['params']
            else:
                kwargs['payload'] = {}
            response = requests.post('{}{}'.format(self.BASE_URL, full_path), data=kwargs['payload'],
                                     headers=self.token)

        if method == 'delete':
            response = requests.delete('{}{}'.format(self.BASE_URL, full_path), headers=self.token)

        res = self._handle_response(response)
        return res

    @staticmethod
    def _handle_response(json_response):
        """
        helper to handle api responses and determine exceptions
        """
        if not str(json_response.status_code).startswith('2'):
            raise JockAPIException(json_response)
        try:
            res = json_response.json()
            return res
        except ValueError:
            raise JockAPIException('Invalid Response: %s' % json_response.text)

    def _throttle_requests(self, func):
        """
        method designed to throttle certain user requests
        """
        pass

    def _get(self, path, api_version=None, **kwargs):
        """
        method for get requests
        """
        return self._request('get', path, api_version, **kwargs)

    def _post(self, path, api_version=None, **kwargs):
        """
        method for post requests
        """
        return self._request('post', path, api_version, **kwargs)

    def _delete(self, path, api_version=None, **kwargs):
        """
        method for delete requests, used exclusively for order deletion
        """
        return self._request('delete', path, api_version, **kwargs)

    def _get_account_bal(self):
        """
        method retreiving user's USD balance
        """
        return round(self._get("balances")['balances'][0]['buying_power'], 2)

    def _get_account(self):
        """
        method retreiving user's account balance
        """
        return self._get("account")['account']

    def get_teams(self, start: int = 0, league: str = None) -> list[Team]:
        """
        provides a list of teams for all or chosen leagues that have team structure.
        displays only the first page, the user can paginate via:
        for i in range(n):
            get_teams(league = x, start = i)

        keyword args:
            start -- default: 0 (first 100 responses)
            league -- default: None (Any), or nba, nfl, mlb, nhl, nascar (lowercase)
        """
        teams = []
        params = {'start': str(start * 100), 'limit': '100'}
        if league is not None:
            params['league'] = league
        res = self._get('teams', params=params)
        print('status: ' + res['status'])
        print('start: ' + str(res['start']))
        print('limit: ' + str(res['limit']))
        print('count: ' + str(res['count']))
        for team in res['teams']:
            teams.append(Team(team))
        return teams

    def get_team(self, team_id: str) -> Team:
        """fetch a specific entity based on their entity id

        Keyword args:
            team_id: required (str) e.g. "team_8fe94ef0d1f0a00e1285301c4092650f"
        """
        team = self._get(f"teams/{team_id}")['team']
        return Team(team)

    def get_entities(self, start: int = 0, limit: int = 100, include_team: bool = True, league: str = None) -> list[Entity]:
        """fetch entities (players of any sport)
        the user will have to paginate i.e.
            for i in range(n):
                get_entities(start=i)
        
        Keyword args:
            :param start: page at which the user wants to start their search, default: 0
            :type start: int

            :param limit: number of entities the user wants to display, default: 100
            :type limit: int


            :param include_team: include team information for the entity, default: False
            :type include_team: bool

            :param league: filter by league, either 'nba', 'nhl, 'nfl', 'pga', 'mlb', 'nascar'
            :type league: str
        """
        params = {}
        entities = []
        if league is not None:
            params['league'] = league
        if include_team:
            params['include'] = 'team'
        params['start'] = start * limit
        params['limit'] = limit
        res = self._get("entities", params=params)
        print('status: ' + res['status'])
        print('start: ' + str(res['start']))
        print('limit: ' + str(res['limit']))
        print('count: ' + str(res['count']))
        for entity in res['entities']:
            entities.append(_case_switch_ent(entity))
        return entities

    def get_entity(self, entity_id: str, include_team: bool = False) -> Entity:
        """fetch a specific entity based on their entity id

        Keyword args:
            :param entity_id: the chosen entity's id, required
            :type entity_id: str

            :param include_team: include team information for the entity, default: False
            :type include_team: bool
        """
        if include_team:
            params = {'include': 'team'}
        else:
            params = {}
        ent = self._get(f"entities/{entity_id}", params=params)['entity']
        return _case_switch_ent(ent)

    def get_games(self, start: int = 0, limit: int = 100, league: str = None) -> list[Game]:
        """provides a list of teams for all or chosen leagues that have team structure
        the user will have to paginate.
        i.e. for i in range(n):
                get_games(start=i)

        keyword args:
            start -- default: 0 (int), which page of games the user wishes to see
            limit -- default: 100, displays chosen number of relevant events
            league -- default: None (Any), or nba, nfl, mlb, nhl, nascar, pga
        """
        params = {}
        response = []
        params['start'] = start * limit
        params['limit'] = limit
        if league is not None:
            params['league'] = league
        res = self._get('games', params=params)
        print('status: ' + res['status'])
        print('start: ' + str(res['start']))
        print('limit: ' + str(res['limit']))
        print('count: ' + str(res['count']))
        for game in res['games']:
            response.append(Game(game))
        return response

    def get_game(self, game_id: str) -> Game:
        """fetch a specific entity based on their entity id

        Keyword args:
        game_id -- required, str, e.g. "game_60bb686586eaf95a5e8dafa3823d89cb"
        """
        return Game(self._get(f"games/{game_id}")['game'])

    def get_game_logs(self, start: int = 0, limit: int = 100, log_id: str = None, entity_id: str = None,
                      game_id: str = None, include_ent: bool = False, include_game: bool = False,
                      include_team: bool = False) -> list[GameLog]:
        """fetch game logs

        Keyword args:
        Optional:
            start -- default: 0 (int), which page of games the user wishes to see
            limit -- default: 100, displays chosen number of most recent logs, max 100
            log_id -- str or list, filter for a specific game log. e.g. "gl_60cde7f973f9e00674785e5e144a802b"
            entity_id -- str, filter all game logs for a specific player, e.g. "en_67c8368a3905f8beee69393ccec854e5"
            game_id -- str, filter all game logs for all players in a specific game,
            e.g. "game_60cde69ee06e791b99ed71e6013fc4a7"
            include_ent -- bool, if True, returns entity information attached to the game log (entity name, team, etc.)
            include_game -- bool, if True, returns game information (game name, teams, status)
            include_team -- bool, if True, returns team information (team name, location, league, etc.)
        """
        params = {}
        include = []
        response = []
        if log_id is not None:
            params['id'] = log_id
        if entity_id is not None:
            params['entity_id'] = entity_id
        if game_id is not None:
            params['game_id'] = game_id
        if include_game:
            include.append('game')  # can be computationally costly for golf events, be careful
        if include_ent:
            include.append('entity')
        if include_team:
            include.append('team')
        if len(include) != 0:
            params['include'] = str(include)
        params['start'] = start * limit
        params['limit'] = limit
        res = self._get("game_logs", params=params)
        print('status: ' + res['status'])
        print('start: ' + str(res['start']))
        print('limit: ' + str(res['limit']))
        print('count: ' + str(res['count']))
        for log in res['game_logs']:
            print(log)
            response.append(GameLog(log))
        return response

    def get_events(self, start: int = 0, limit: int = 25, league: str = None, include_sims: bool = False) \
            -> list[Event]:
        """Populates event objects with recent and upcoming events

        Keyword args:
            start -- default: 0 (int), which page of games the user wishes to see
            limit -- default: 25, displays chosen number recent and upcoming events
            league -- default: None (Any), or nba, nfl, mlb, nhl, nascar, pga
            include_sims -- default: False (bool), if True, will return Horse Sims for test use
        """
        list_events = []
        data = {'start': str(start * limit), 'limit': limit}
        if league is not None:
            data['league'] = league
        res = self._get('events', data=data)
        for event in res['events']:
            if event['league'] != 'simulated_horse_racing':
                list_events.append(Event(event))
            elif include_sims:
                list_events.append(Event(event))
        return list_events

    def get_event(self, event_id: str) -> Event:
        """fetch a particular event, by default includes games, payouts and tradeables. This is easier than
        pulling payouts, tradeables, and games separately

        Keyword args:
        Required:
            event_id -- str, the event_id for your chosen event, e.g. evt_60dbec530d2197a973c5dddcf6f65e12
        """
        params = {}
        include = ['tradeables.entity', 'games', 'payouts']
        params['include'] = str(include)
        res = self._get(f"events/{event_id}", params=params)
        return Event(res['event'])

    def get_event_payouts(self, event_id: str) -> json:  # should this be appended to the event object itself?
        """get payouts for each rank of an event

        Keyword args:
        Required:
            event_id -- str, the event_id for your chosen event, e.g. evt_60dbec530d2197a973c5dddcf6f65e12
        """
        return self._get(f"events/{event_id}/payouts")

    def get_event_games(self, event_id: str) -> list[Game]:
        """get all games in an event

        Keyword args:
        Required:
            event_id -- str, the event_id for your chosen event, e.g. evt_60dbec530d2197a973c5dddcf6f65e12
        """
        games = []
        res = self._get(f"events/{event_id}/games")
        for game in res['games']:
            games.append(Game(game))
        return games

    def get_event_tradeables(self, event_id: str) -> list[Tradeable]:
        """get all tradeables in an event

        Keyword args:
        Required:
            event_id -- str, the event_id for your chosen event, e.g. evt_60dbec530d2197a973c5dddcf6f65e12
        """
        res = self._get(f"events/{event_id}/tradeables")
        tradeables = []
        for tdbl in res['tradeables']:
            tradeables.append(Tradeable(tdbl))
        return tradeables

    def get_entries(self, start: int = 0, limit: int = 10, include_payouts: bool = False,
                    include_tradeables: bool = False) -> list[Entry]:
        """obtain information about events a user has entered

        Keyword args:
        Optional:
            start -- default: 0 (int), which page of games the user wishes to see
            limit -- default: 10 (int), how many recent entries the user would like to see
            include_payouts -- default: False, include the payouts for each rank of an event
            include_tradeables -- default: False, if True, includes all tradeables participating in the event
            (id, fpts, event rank, prices, stats, etc.)
        """
        params = {'start': start * limit, 'limit': limit}
        include = ['event']
        if include_payouts:
            include.append('payouts')
        if include_tradeables:
            include.append('payouts.tradeable')
        params['include'] = str(include)
        print(params)
        response_list = []
        res = self._get("entries", params=params)
        print('status: ' + res['status'])
        print('start: ' + str(res['start']))
        print('limit: ' + str(res['limit']))
        print('count: ' + str(res['count']))
        for entry in res['entries']:
            response_list.append(entry)

            # potential todo: add if loop to display only the most recent few events, and calculate profit and loss
            #  for the event (including fees, $ invested, etc.)
        return response_list

    def calculate_pnl(self, entry_id: str):
        pass

    def get_entry(self, entry_id: str, include_event=False, include_payouts=False,
                  include_tradeables=False) -> Entry:
        """obtain information about events a user has entered

        Keyword args:
        Optional:
            qty -- default: 50, number of events the user wishes to display
            include_event -- default: False, include event information
            include_payouts -- default: False, include the payouts for each rank of an event (only displayed post-event)
            include_tradeables -- default: False, if True, includes all tradeables participating in the event
            (only displayed post-event)
            (provides information about positions/payouts and their underlying tradeables)
        """
        params = {}
        include = []
        if include_event:
            include.append('event')
        if include_payouts:
            include.append('payouts')
        if include_tradeables:
            include.append('payouts.tradeable')
        if len(include) != 0:
            params['include'] = str(include)
        print(params)
        entry = self._get(f"entries/{entry_id}", params=params)
        print(entry)
        return Entry(entry['entry'])

    def create_entry(self, event_id: str) -> json:
        """create an entry to an event given an event_id e.g. evt_60dbec530d2197a973c5dddcf6f65e12
        """
        return self._post(f"entries", data={'event_id': event_id})

    _order_cache = []

    # @staticmethod
    # def _order_counter(self, func):
    #     def _helper(*args, **kwargs):
    #         pass
    #     pass

    def place_order(self, id: str, price: int, side: str = 'buy', phase: str = 'ipo', qty: int = 1, **kwargs) -> json:
        """
        TODO: exception handling, if response to an order is TOO MANY ORDERS, add the order back to the queue
        add functionality for 'market' orders
        add functionality for take_profit/stop_loss
        add functionality for fill or kill/good until
        add functionality to allow user to place order at JockMkt estimated price
        get event status and auto choose phase

        keyword args:
            id -- your player's tradeable id, which can be obtained by...
            side -- default: buy, buy or sell (sell only available during live trading
            phase -- default: ipo or live
            qty -- default: 1, number of shares you wish to purchase
            price -- bid/ask per share. Do not include your fees.

        optional:
            order_size -- input a total amount to spend and calculate desired number of shares automatically
        """
        if price > 25:
            price = 25

        if 'order_size' in kwargs:
            size = kwargs.get('order_size', 0)
            qty = size // price

        price = "{:.2f}".format(price)

        order = {'tradeable_id': id, 'side': side, 'type': 'limit', 'phase': phase, 'quantity': str(qty),
                 'limit_price': str(price)}
        order_response = self._post('orders', data=order)
        print(order_response)

        return order_response

    # NOTE: the docs for order object > status contain 'outbid' twice

    def get_orders(self, start=0, limit: int = 100, event_id: str = None, active: bool = False,
                   updated_after: int = None):
        """get all of a user's orders
        Keyword args:
        optional:
            qty -- int, number of orders the user wishes to display
            event_id -- str, orders relating to a specific event
            active -- bool, include only orders marked (created or accepted, not filled, canceled, outbid, or expired)
            updated_after -- int, ms timestamp 13 digits, includes orders only after the timestamp

        # TODO Order counter, pass tradeables with open orders
        """
        params = {'start': start * limit, 'limit': limit}
        orders = []
        if event_id is not None:
            params['event_id'] = str(event_id)
        if active:
            params['active'] = str(True)
        if updated_after is not None:
            params['updated_after'] = str(updated_after)
        orders_response = self._get('orders', params=params)
        print('status: ' + orders_response['status'])
        print('start: ' + str(orders_response['start']))
        print('limit: ' + str(orders_response['limit']))
        print('count: ' + str(orders_response['count']))
        for order in orders_response['orders']:
            orders.append(Order(order))
        return orders

    def get_order(self, order_id: str) -> Order:
        """get information about a specific order

        Keyword args:
        required:
            order_id -- str, e.g. ord_601b5ad6538ec34875ee1687c4a657f8
        """
        return Order(self._get(f"orders/{order_id}")['order'])

    def delete_order(self, order_id: str) -> json:
        """delete a specific order

        Keyword args:
        required:
            order_id -- str, e.g. ord_601b5ad6538ec34875ee1687c4a657f8
        """
        deletion_res = self._delete(f"orders/{order_id}")
        if deletion_res['status'] == 'success':
            print('order successfully canceled')
        print(deletion_res)
        return deletion_res

    def get_positions(self) -> list[Position]:
        """returns a user's open positions in all current events
        """
        positions = []
        positions_res = self._get("positions")
        print('status: ' + positions_res['status'])
        print('count: ' + str(positions_res['count']))
        for position in positions_res['positions']:
            positions.append(Position(position))
        return positions

    def get_account_activity(self, start: int = 0, limit: int = 100) -> list[AccountActivity]:
        """
        returns a user's most recent account activity
        :param start: the page of account activity which the user wants to display
        :type start: int
        :default start: 0
        :param limit: the quantitiy of account activity objects the user wants to see, max = 100
        :type limit: int
        :default limit: 100
        """
        params = {'start': str(start * limit), 'limit': limit}
        activity_res = self._get('account/activity', params=params)
        acct_activity = []
        for aact in activity_res['activity']:
            acct_activity.append(AccountActivity(aact))
        return acct_activity

    @staticmethod
    def fetch_all(func, pages: int) -> object:
        pass
